src/main/java/com/iexec/common/utils/ContextualLockRunner.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2021 IEXEC BLOCKCHAIN TECH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.iexec.common.utils;
+
+import net.jodah.expiringmap.ExpiringMap;
+
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Provides a way to avoid parallel runs of an action.
+ * Locks are based on a key, so that two actions with the same key can't be run at the same time
+ * but two actions with different keys can be run at the same time.
+ * <br>
+ * To avoid memory-leaks, entries expire after a given time.
+ * Default is 1 hour, but it can be set at {@link ContextualLockRunner} object creation.
+ *
+ * @param <K> Type of the key.
+ */
+public class ContextualLockRunner<K> {
+    private final ConcurrentMap<K, Object> locks;
+
+    public ContextualLockRunner() {
+        this(1, TimeUnit.HOURS);
+    }
+
+    public ContextualLockRunner(long expirationDuration, TimeUnit unit) {
+        this.locks = ExpiringMap.builder()
+                .expiration(expirationDuration, unit)
+                .build();
+    }
+
+    public void runWithLock(K key, Runnable action) {
+        synchronized (getLock(key)) {
+            action.run();
+        }
+    }
+
+    public void acceptWithLock(K key, Consumer<K> action) {
+        synchronized (getLock(key)) {
+            action.accept(key);
+        }
+    }
+
+    public <R> R getWithLock(K key, Supplier<R> action) {
+        synchronized (getLock(key)) {
+            return action.get();
+        }
+    }
+
+    public <R> R applyWithLock(K key, Function<K, R> action) {
```

Comment:
It's a little confusing that `key` is both used to lock and as a parameter of the action. We can maybe add a third argument to the method (`..., T actionArgument`) but I'm afraid it will make the method too complicated to read.

Reply:
IMHO, it's better as is. You'll generally want to execute an action on a given key, e.g. when you'll want to run some action and lock it based on a `chainTaskId`, it'll probably be the argument passed to the action.